Cabal/Distribution/Types/VersionRange.hs

```diff
@@ -340,34 +358,94 @@ withinRange v = foldVersionRange
 -- Wildcard range utilities
 --
 
--- | @since 2.2
-wildcardUpperBound :: Version -> Version
-wildcardUpperBound = alterVersion $
-    \lowerBound -> init lowerBound ++ [last lowerBound + 1]
-
 isWildcardRange :: Version -> Version -> Bool
 isWildcardRange ver1 ver2 = check (versionNumbers ver1) (versionNumbers ver2)
   where check (n:[]) (m:[]) | n+1 == m = True
         check (n:ns) (m:ms) | n   == m = check ns ms
         check _      _                 = False
 
--- | Compute next greater major version to be used as upper bound
---
--- Example: @0.4.1@ produces the version @0.5@ which then can be used
--- to construct a range @>= 0.4.1 && < 0.5@
---
--- @since 2.2
-majorUpperBound :: Version -> Version
-majorUpperBound = alterVersion $ \numbers -> case numbers of
-    []        -> [0,1] -- should not happen
-    [m1]      -> [m1,1] -- e.g. version '1'
-    (m1:m2:_) -> [m1,m2+1]
-
 -------------------------------------------------------------------------------
 -- Parsec & Pretty
 -------------------------------------------------------------------------------
 
-instance Pretty VersionRange where
+class Ord v => IsVersion v where
+  dispWild :: v -> Disp.Doc
+  noVersion :: VersionRangeOver v
+  majorUpperBound :: v -> v
+  wildcardUpperBound :: v -> v
+  verOrWild :: CabalParsing m => m (Bool, v)
+  parseWildcardRange :: Parse.ReadP r (VersionRangeOver v)
+
+instance IsVersion Version where
+  dispWild ver = Disp.hcat (Disp.punctuate (Disp.char '.') (map Disp.int $ versionNumbers ver))
+                 <<>> Disp.text ".*"
+
+  noVersion = IntersectVersionRanges (LaterVersion v) (EarlierVersion v)
+    where v = mkVersion [1]
+
+
+  -- | Compute next greater major version to be used as upper bound
+  --
+  -- Example: @0.4.1@ produces the version @0.5@ which then can be used
+  -- to construct a range @>= 0.4.1 && < 0.5@
+  --
+  -- @since 2.2
+  majorUpperBound = alterVersion $ \numbers -> case numbers of
```

Comment:
because instance member haddocks are IMHO hidden quite well.

Reply:
also what will `^>=` mean for `PkgConfigVersion`? Should that be disallowed? structurally or in `cabal check`?

Reply:
Well,  `^>=` assumes a well established semantic versioning policy/convention. This doesn't  exist for  `pkg-config(1)` versions. The meaning of `^>=` for pkgconfig is thus IMO not well-defined, and we should discourage its use until we come up with a good one. So at the very least a `cabal check` warning/error is in order.